### PR TITLE
Enhance error messages for duplicate constructors

### LINKED ISSUE #47272: Diagnostic for conflicting constructor declarations refer to the conflicting member as '.ctor'
**Version Used**: 3.7.0-6.20418.4 (9b878f99)

**Steps to Reproduce**:
Define a type with conflicting constructor signatures:
```cs
public class C
{
    public C() { }
    public C() { }
}
```
**Expected Behavior**:
An error which says "Type 'C' already defines a constructor with the same parameter types"

**Actual Behavior**:
error CS0111: Type 'C' already defines a member called '.ctor' with the same parameter types

This compiler diagnostic seems to have been referring to the conflicting member as '.ctor' for a while now. Before that, I believe it would have referred to the conflicting member as 'C'.
Ideally, I think the error message should directly say that there are conflicting *constructors*, since constructors do not naturally have their own name. Regardless, referring to the member as '.ctor' is confusing and should be changed.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #47272

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 1]
```diff
@@ -1728,8 +1728,21 @@ private void ReportMethodSignatureCollision(DiagnosticBag diagnostics, SourceMem
...
-            diagnostics.Add(ErrorCode.ERR_MemberAlreadyExists, method1.Locations[0], methodName, this);
+            if (methodName == WellKnownMemberNames.InstanceConstructorName)
+            {
+                // Type '{0}' already defines a constructor with the same parameter types
+                diagnostics.Add(ErrorCode.ERR_ConstructorAlreadyExists, method1.Locations[0], this);
+            }
+            else if (methodName == WellKnownMemberNames.StaticConstructorName)
+            {
+                // Type '{0}' already defines a static constructor
+                diagnostics.Add(ErrorCode.ERR_StaticConstructorAlreadyExists, method1.Locations[0], this);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1739]: Are we testing this case?

**Comment by cston** [comment]
It's not clear to me whether we should fix this case, since there are cases other than constructors where the compiler reports unexpected names - for instance, indexers and operators: [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmAAgGEBYAKAG9yCaCcBmOgRgDZiCB7AB0Qhg4QEA1AAoiBCEnbAAlAQp0A7BIDcBAL7VaDZm3Hde/QaPGTpchTmUQ1msrTqMOwAFZwAxjAIwAFmigA2mgAdl4hACZwAB4AuvIEAOZwXpbKwQCuGBi2Glo0Os5unt5+gSFhwZGx8UkpSgQZWTl2dkA==).

If we should fix the constructor case, perhaps we should follow the pre-Roslyn compiler and use the existing error message with the type name rather than ".ctor".

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1739]: @cston I'm not sure. Waiting for the build to see if there are build failures related to this or not. If there are not, I'll add new test for this.

**Comment by Youssef1313** [comment]
> 
> 
> It's not clear to me whether we should fix this case, since there are cases other than constructors where the compiler reports unexpected names - for instance, indexers and operators: [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmAAgGEBYAKAG9yCaCcBmOgRgDZiCB7AB0Qhg4QEA1AAoiBCEnbAAlAQp0A7BIDcBAL7VaDZm3Hde/QaPGTpchTmUQ1msrTqMOwAFZwAxjAIwAFmigA2mgAdl4hACZwAB4AuvIEAOZwXpbKwQCuGBi2Glo0Os5unt5+gSFhwZGx8UkpSgQZWTl2dkA==).
> 
> If we want to fix the constructor case, perhaps we should follow the pre-Roslyn compiler and use the existing error message with the type name rather than ".ctor".

@cston Should I then update the logic to handle indexers and operators as well?
But I don't think the chain of the if conditions will be good.

**Comment by cston** [comment]
> Should I then update the logic to handle indexers and operators as well?
> But I don't think the chain of the if conditions will be good.

I don't think it is worth handling each of those cases.

I think it would be sufficient to clean up the constructor cases by using the existing error, `ERR_MemberAlreadyExists`, but with the type name rather than the method name: "Type 'C' already contains a member called 'C' with the same parameter types". That would match the compiler behavior pre-Roslyn.

**Comment by Youssef1313** [comment]
@cston What should be the correct way to retrieve the value of `methodName` in that case?

**Comment by cston** [comment]
> What should be the correct way to retrieve the value of `methodName` in that case?

Perhaps: `var methodName = ... method1.IsConstructor() ? this.Name : method1.Name;`

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 2]
```diff
@@ -1727,7 +1727,8 @@ private void ReportMethodSignatureCollision(DiagnosticBag diagnostics, SourceMem
             // Special case: if there are two destructors, use the destructor syntax instead of "Finalize"
             var methodName = (method1.MethodKind == MethodKind.Destructor && method2.MethodKind == MethodKind.Destructor) ?
                 "~" + this.Name :
-                method1.Name;
+                (method1.IsConstructor() ? this.Name : method1.Name);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1730]: Please add a test for static constructors.

**Comment by 333fred** [comment]
Thanks @Youssef1313!

**Pull request merged**